### PR TITLE
gstreamer, meson, glib

### PR DESCRIPTION
**Pull request by rhubert**
- a few changes to the meson class
- updated glib to latest version
- added gstreamer recipes

### CODE REVIEW

File: classes/meson.yaml [hunk 1]
```diff
@@ -52,9 +56,14 @@ buildScript: |
...
             declare -a MESON_OPTIONS=( "$1" "$PWD")
             if [[ -e $BOB_CWD/cross_file.txt ]] ; then
+
+                if [[ "${MESON_FILE:-undefined}" != "undefined" ]]; then
+                    cat ${BOB_TOOL_PATHS['target-toolchain']}/${MESON_FILE} >> ../$cross_native
+                fi
                 MESON_OPTIONS+=("--cross-file" "$BOB_CWD/cross_file.txt")
                 MESON_OPTIONS+=("-Ddefault_library=${BASEMENT_LIBS:-shared}")
             else
+                MESON_OPTIONS+=("--native-file" "$BOB_CWD/native_file.txt")
```
Comment (jkloetzke) [inline classes/meson.yaml:66]: Why is this needed for host builds?

File: recipes/multimedia/gstreamer.yaml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    scm: url
+    url: https://gstreamer.freedesktop.org/src/gstreamer/gstreamer-${PKG_VERSION}.tar.xz
+    digestSHA1: 2a417e7c5eb184b705e0964bb63f83f1498bdf84
+
+buildVars: [PKG_VERSION]
+buildTools: [flex, bison]
+buildScript: |
+   export BISON_PKGDATADIR="${BOB_TOOL_PATHS[bison]}/../share/bison"
+   mesonAddTool bison "${BOB_TOOL_PATHS[bison]}/bison-real"
+   mesonAddTool flex "${BOB_TOOL_PATHS[flex]}/flex"
```
Comment (jkloetzke) [inline recipes/multimedia/gstreamer.yaml:23]: That looks odd. According to the documentation tools are searched in `$PATH` automatically. Also, why is it necessary to re-do the bison wrapper script here?

**Review by jkloetzke — COMMENTED** [review]
Most of the changes left me wondering why they are necessary. I merged the long running PR for the Windows cmake changes that included some tests for meson too (tests/linux). They are working for native and cross compiling so far...

File: classes/meson.yaml [hunk 3]
```diff
@@ -19,30 +19,34 @@ buildScript: |
...
+    fi
 
-        cat >cross_file.txt <<EOF
+    cat >$cross_native <<EOF
     [binaries]
     c     = '$CC'
     cpp   = '$CXX'
     ar    = '$AR'
     strip = '$STRIP'
+    pkgconfig = '${BOB_TOOL_PATHS["pkg-config"]}/pkg-config'
```
[the remark this answers is not in the document]
Comment (rhubert) [reply classes/meson.yaml:37]: without this meson building gstreamer can't find pkg-config and fails later when looking for `glib`

```
Checking whether type "struct itimerspec" has members "it_interval", "it_value" : YES 
Pkg-config binary for MachineChoice.HOST is not cached.
None of 'PKG_CONFIG' are defined in the environment, not changing global flags.
Pkg-config binary missing from cross or native file, or env var undefined.
Found Pkg-config: NO
Pkg-config binary for machine MachineChoice.HOST not found. Giving up.
[....]
Run-time dependency glib-2.0 found: NO (tried pkgconfig and cmake)
```

[log.tar.gz](https://github.com/BobBuildTool/basement/files/5686747/log.tar.gz)

File: classes/meson.yaml [hunk 4]
```diff
@@ -52,9 +56,14 @@ buildScript: |
         if [[ ! -e .meson-done ]] ; then
             declare -a MESON_OPTIONS=( "$1" "$PWD")
             if [[ -e $BOB_CWD/cross_file.txt ]] ; then
+
+                if [[ "${MESON_FILE:-undefined}" != "undefined" ]]; then
+                    cat ${BOB_TOOL_PATHS['target-toolchain']}/${MESON_FILE} >> ../$cross_native
+                fi
```
[the remark this answers is not in the document]
Comment (rhubert) [reply classes/meson.yaml:62]: I thought about this but couldn't find a easy way to translate this. (it was easier to add a host-file...) Will look into the translation again. ATM the `cpu` information is not available?

File: classes/meson.yaml (hunk 1 above)
Comment (rhubert) [reply classes/meson.yaml:66]: I don't think it's necessary but with this the same file as for cross compiling can be used to tell meson about the system. For me this makes thinks more clear and the class easier to understand..

File: recipes/multimedia/gstreamer.yaml (hunk 2 above)
Comment (rhubert) [reply recipes/multimedia/gstreamer.yaml:23]: Without this `gstreamer` build fails cause `bison` and `flex` can't be found. Directly executing the bison wrapper also failed with `exec format error`.

Comment (rhubert) [reply recipes/multimedia/gstreamer.yaml:23]: would it be possible to extend bob in order to remove the wrapper by using 

```
provideVars:
 BISON_PKGDATADIR: "$SOME_COOL_BOB_VAR/../share/bison"
```

in the `bison.yaml`?

Comment (jkloetzke) [reply recipes/multimedia/gstreamer.yaml:23]: No, that won't work because all variables are computed independent of the build location.

I'm currently testing your changes. The `exec format error` is caused by a missing `#!/bin/sh` header of the wrapper script. Unfortunately this leads to a rebuild of *everything*. :sob:

**Comment by jkloetzke** [comment]
I've tried to built the recipes here but found numerous issues that surfaced when building in the sandbox. I took the freedom to push the result onto your branch. I have not run any of the results. I did just built them...

**Comment by rhubert** [comment]
Thanks!
But for some reasons the cmake build is now broken :(  There are many undefined references to libssl:

```
x86_64-bob_compat-linux-gnu/1/workspace/usr/lib/libcurl.a(libcurl_la-md5.o): In function `Curl_md5it':
md5.c:(.text+0x17): undefined reference to `MD5_Init'
```

Looks like [this patch](https://github.com/BobBuildTool/basement/blob/master/recipes/devel/cmake/0001-fix-curl-lib-deps.patch) is no longer working? Any ideas?

**Comment by rhubert** [comment]
The pkgconfig files were broken...now cmake builds again. (the remaining build is still pending - rebuilding after changing the install class take some time...)
